### ISSUE TITLE
Make `h` app able to connect to newer ES instance

### ISSUE THREAD

**Issue by lyzadanger** [hypothesis/product-backlog#631]
Before we can read to or write from a new ES instance’s index, the app has to _connect to the ES instance_. At present the connection to our existing, legacy ES instance is created during Pyramid bootstrapping in the `h.search.__init__` module.

We need to be able to establish a connection to both the current v1.5 and the newer v6.x within the `h` app. This is a prerequisite to being able to initialize the ES index and subsequently to be able to use it for any sort of read/write.

At time of writing it is still TBD whether implementation of this will result in multiple `h.search.client.Client` instances or whether something internal in the “client” wrapper will delegate between the two instances. My hunch is though the latter sounds more elegant, the former _may_ be more practicable—but I’m happy to be proven wrong.

At present:

* `request.es` is a reference to an instance of `h.search.client.Client` with a connection to an ES 1.5 instance at the port indicated by `settings.es_host` (see: `conf.development-app.ini` or `conf.app.ini` (or overridden by the `ELASTICSEARCH_HOST` env)
* As the `Client` that becomes `request.es` is created during Pyramid bootstrapping, there is no `request` yet available to check feature flags against. We discussed as a team having the second instance connection be based on the presence of a setting versus a feature flag.

Additional things to consider:

* `h.search.client.get_client` also involves AWS authentication if certain settings are present—it’s not going to be feasible (I don’t think?) testing this on local instances. Something to keep in mind.
* Several of the `h.search.*` functions and constructors accept an `es` parameter (i.e. a client), but in reality this is always (so far as I have seen) `request.es`. The reading (i.e. search) stuff always uses `request.es` anyway. Depending on the approach for this work, it may (or may not) be necessary to address that and decide whether `es` is really a valid parameter for those functions or if they should just take a `request`. Note that at present the `h.search.index.delete` function is the only one I see that doesn’t have access to `request` already.

**Comment by lyzadanger** [comment]
Useful note: Multiple versions of the [`elasticsearch`](https://elasticsearch-py.readthedocs.io/en/master/) library can be installed by using `elasticsearch2`, `elasticsearch6`, e.g.

**Comment by seanh** [comment]
[Relevant suggestion in Slack](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1525345236000035?thread_ts=1525343480.000340&cid=C1MA4E9B9):

- Add `ELASTICSEARCH_URL` env var to qa and prod (if we don't have the actual value for it yet, just set it to something) (It looks like this is being done as part of [this card](https://github.com/hypothesis/product-backlog/issues/604))

- Add `es.url` default to `development-app.ini` for dev envs and tests (this can point to the actual working Es 6.2 that we already have in dev envs and on Travis and Jenkins, Jenkins will need to set an env var for it since it uses a randomly chosen port)

- Add `ELASTICSEARCH_URL` as a _required_ setting in the code. It should error on startup if the required setting is missing. h doesn't yet have the concept of a required setting that errors on startup if missing, but the old lti app had it (with very similar config settings code to h): https://github.com/hypothesis/lti/blob/0cf85fdfff4e4fa92d03a1ee5d2aaea430f9cdfb/lti/config/settings.py#L14

- So now we'll have the new required elasticsearch url setting, and can go crazy writing code that assumes its presence. (And if we don't want some of that code to execute in prod yet, make those paths conditional on a different, temporary env var for that specific purpose)

- I think the alternative approach of making the setting optional now and required later is also reasonable, but we should as much as possible write code that assumes its presence and not introduce "if not set then X" complexity that won't be needed in the long term, except where really necessary and temporary. and we should make sure to do the follow-up work afterwards of making the setting required (including implementing the concept of a required setting in h) and removing any now-dead code that deals with the setting not being there

**Comment by seanh** [comment]
Closing in favour of https://github.com/hypothesis/product-backlog/issues/648
